api/src/OpenTelemetry/Internal/Logging/Types.hs

```diff
@@ -43,8 +53,76 @@ data Logger = Logger
 
 {- | This is a data type that can represent logs from various sources: application log files, machine generated events, system logs, etc. [Specification outlined here.](https://opentelemetry.io/docs/specs/otel/logs/data-model/)
 Existing log formats can be unambiguously mapped to this data type. Reverse mapping from this data type is also possible to the extent that the target log format has equivalent capabilities.
+Uses an IORef under the hood to allow mutability.
+-}
+data LogRecord a = LogRecord Logger (IORef (ImmutableLogRecord a))
+
+
+mkLogRecord :: Logger -> ImmutableLogRecord body -> IO (LogRecord body)
+mkLogRecord l = fmap (LogRecord l) . newIORef
+
+
+{- | This is a typeclass representing @LogRecord@s that can be read from.
+
+A function receiving this as an argument MUST be able to access all the information added to the LogRecord. It MUST also be able to access the Instrumentation Scope and Resource information (implicitly) associated with the LogRecord.
+
+The trace context fields MUST be populated from the resolved Context (either the explicitly passed Context or the current Context) when emitted.
+
+Counts for attributes due to collection limits MUST be available for exporters to report as described in the transformation to non-OTLP formats specification.
 -}
-data LogRecord body = LogRecord
+class ReadableLogRecord r where
+  -- | Reads the current state of the @LogRecord@ from its internal @IORef@. The implementation mirrors @readIORef@.
+  readLogRecord :: r a -> IO (ImmutableLogRecord a)
+
+
+  -- | Reads the @InstrumentationScope@ from the @Logger@ that emitted the @LogRecord@
+  readLogRecordInstrumentationScope :: r a -> InstrumentationLibrary
+
+
+  -- | Reads the @Resource@ from the @LoggerProvider@ that emitted the @LogRecord@
+  readLogRecordResource :: r a -> MaterializedResources
+
+
+{- | This is a typeclass representing @LogRecord@s that can be read from or written to. All @ReadWriteLogRecord@s are @ReadableLogRecord@s.
+
+A function receiving this as an argument MUST additionally be able to modify the following information added to the LogRecord:
+
+- Timestamp
+- ObservedTimestamp
+- SeverityText
+- SeverityNumber
+- Body
+- Attributes (addition, modification, removal)
+- TraceId
+- SpanId
+- TraceFlags
+-}
+class (ReadableLogRecord r) => ReadWriteLogRecord r where
+  -- | Reads the attribute limits from the @LoggerProvider@ that emitted the @LogRecord@. These are needed to add more attributes.
+  readLogRecordAttributeLimits :: r a -> AttributeLimits
+
+
+  -- | Modifies the @LogRecord@ using its internal @IORef@. This is lazy and is not an atomic operation. The implementation mirrors @modifyIORef@.
+  modifyLogRecord :: r a -> (ImmutableLogRecord a -> ImmutableLogRecord a) -> IO ()
```

Comment:
There is currently no strict version of `modifyLogRecord`. It might be better to make `modifyLogRecord` by default  strict and export no lazy version, but I could also write both `modifyLogRecord` and `modifyLogRecord'`.